### PR TITLE
feat: List available channel types

### LINKED ISSUE #979: kn channel list-types
### Feature request

List available channel types in cluster which can be used for creating channels.

### Use case
As a user I want to list available channel types in my cluster which can be used to create channels.

### UI Example
```
kn channel list-types
TYPE             GROUP                   VERSION
InMemoryChannel  messaging.knative.dev   v1abeta1
KafkaChannel     messaging.knative.dev   v1alpha1
```

### Implementation notes:
- The implementation could follow `kn source list-types` approach to query CRDs using dynamic client,
  filtered by label `messaging.knative.dev/subscribable=true`.

/kind proposal

### PR DESCRIPTION
**Pull request by Kaustubh-pande**
## Description

List available channel types in the cluster which can be used for creating channels.
## Changes

Add channel list-types

## Reference

Fixes #979

### CODE REVIEW

**Review by maximilien — APPROVED** [review]
Thanks for contribution. Will do a full review. Let’s get the tests running first.

/ok-to-test

File: docs/cmd/kn_channel.md
Comment (rhuss) [inline docs/cmd/kn_channel.md:34]: While just thinking about this, why not just `kn channel types` ? I know this violates the <noun> <verb> schema, but it looks imo nice (same for `kn source list-types`). At the end "list-types" is *also* not a verb but an artificial word construct.
wdyt @navidshaikh @maximilien ?

File: pkg/dynamic/client_test.go
Comment (rhuss) [inline pkg/dynamic/client_test.go:29]: I wonder whether we should already go to `v1` API (and then change eventing also to v1 in a subsequent PR).

**Review by rhuss — COMMENTED** [review]
Looks fine to me, but would love to have some review from @navidshaikh , too, as he did the source list-types (and this here is quite similar). 

My only suggestion would be whether we might want reconsider the `list-types` name to simplify to `types` (although this would break our _noun verb_ schema) But see the comment below.

File: docs/cmd/kn_channel.md
Comment (navidshaikh) [reply docs/cmd/kn_channel.md:34]: I liked that. I'd keep it `list-types` for this PR and change for source and channel together in a subsequent PR (keeping list-types as an alias for source). cc @Kaustubh-pande

File: lib/test/channel.go
Comment (navidshaikh) [reply lib/test/channel.go:57]: We should add the the comment there as its exported function @Kaustubh-pande

File: pkg/dynamic/client.go
Comment (navidshaikh) [inline pkg/dynamic/client.go:45]: Wherever applicable, we should refer these values from vendored eventing code, good candidate for a subsequent PR.

File: pkg/dynamic/client.go
Comment (navidshaikh) [inline pkg/dynamic/client.go:70]:
```suggestion
	ListChannelsUsingGVKs(*[]schema.GroupVersionKind, ...WithType) (*unstructured.UnstructuredList, error)
```

File: pkg/dynamic/client.go
Comment (navidshaikh) [inline pkg/dynamic/client.go:232]: :+1: this should help filtering when we plan `kn channel list` (like `kn source list`) and operate in restricted access environment

File: pkg/dynamic/client_test.go
Comment (navidshaikh) [reply pkg/dynamic/client_test.go:29]: lets keep it at v1beta1 to also support working with earlier eventing installs?

File: pkg/kn/commands/channel/flags.go
Comment (navidshaikh) [inline pkg/kn/commands/channel/flags.go:132]:
```suggestion
		{Name: "Type", Type: "string", Description: "Kind / Type of the channel", Priority: 1},
```

File: pkg/kn/commands/channel/flags.go
Comment (navidshaikh) [inline pkg/kn/commands/channel/flags.go:134]: We should have `Name` and `Description` fields here instead of `Grop` and `Version`.


```
➜  client git:(channel-list-types) kn channel list-types
TYPE              GROUP                   VERSION
InMemoryChannel   messaging.knative.dev   v1beta1
KafkaChannel      messaging.knative.dev   v1beta1
```
vs
```
➜  client git:(channel-list-types) kn source list-types
TYPE              NAME                                   DESCRIPTION
ApiServerSource   apiserversources.sources.knative.dev   Watch and send Kubernetes API events to addressable
ContainerSource   containersources.sources.knative.dev   Generate events by Container image and send to addressable
KafkaSource       kafkasources.sources.knative.dev       Route events from Apache Kafka Server to addressable
PingSource        pingsources.sources.knative.dev        Send periodically ping events to addressable
SinkBinding       sinkbindings.sources.knative.dev       Binding for connecting a PodSpecable to addressable
```

File: test/e2e/channels_test.go
Comment (navidshaikh) [inline test/e2e/channels_test.go:113]: needs an update according to suggestion in the column fields

**Review by navidshaikh — CHANGES_REQUESTED** [review]
neat! There are a few suggestion as to which fields to print in the columns, we can remove Group and Version columns, and replace them with `NAME` and `DESCRIPTION` to align with `kn source list-types`. Added a couple of suggestions for follow up PR, with the code added in this PR, we can  `kn channel list` next.

File: pkg/messaging/v1beta1/client.go
Comment (daisy-ycguo) [inline pkg/messaging/v1beta1/client.go:61]: According to the CRDs defined in knative messaging, `Channel` is another channel type ( although it has no specific implementation ) which has the label `messaging.knative.dev/subscribable: "true"` too. Here you don't regard it as a kind of build-in `Channel` type. I wonder if it will bring any inconsistency when `listBuiltInChannelTypes` is invoked.

File: pkg/kn/commands/channel/flags.go
Comment (daisy-ycguo) [inline pkg/kn/commands/channel/flags.go:178]: Here you exclude `channels.messaging.knative.dev` manually. But it will bring inconsistency when you call `kn channel list-types -o name` or `kn channel list-types -o yaml`. Here is the test result in my environment:

```
$ ./kn channel list-types -o name
customresourcedefinition.apiextensions.k8s.io/channels.messaging.knative.dev
customresourcedefinition.apiextensions.k8s.io/inmemorychannels.messaging.knative.dev
customresourcedefinition.apiextensions.k8s.io/kafkachannels.messaging.knative.dev
$ ./kn channel list-types
TYPE              GROUP                   VERSION
InMemoryChannel   messaging.knative.dev   v1beta1
KafkaChannel      messaging.knative.dev   v1beta1
```

Please address the issue. If you are not able to remove `customresourcedefinition.apiextensions.k8s.io/channels.messaging.knative.dev` from the output of `kn channel list-types -o name`, maybe you have to add this item back.

**Review by daisy-ycguo — CHANGES_REQUESTED** [review]
Thank you for your contribution, @Kaustubh-pande. I have a few comments. Would you address them and upload a new version ?

File: pkg/kn/commands/channel/list_types.go
Comment (navidshaikh) [inline pkg/kn/commands/channel/list_types.go:65]:
```suggestion
				return fmt.Errorf("no channels found on the backend, please verify the installation")
```

**Review by navidshaikh — APPROVED** [review]
lgtm, just a minor nit for the error message and please add the CHANGELOG entry ?

**Review by daisy-ycguo — APPROVED** [review]
LGTM. 
After tests pass, I'm OK to merge it. Thank you for your contribution.

**Review by navidshaikh — APPROVED** [review]
/lgtm
/approve

**Comment by rhuss** [comment]
```
   ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
        🦆 kn service update svc3 --annotation alpha=direwolf --annotation brave- --namespace kne2etests7
        ┃ Updating Service 'svc3' in namespace 'kne2etests7':
        ┃ 
        ┃   0.045s The Configuration is still working to reflect the latest desired specification.
        ┃   0.404s Ingress reconciliation failed
        ┃ 
        🔥 Error: ReconcileIngressFailed: Ingress reconciliation failed
        🔥 Run 'kn --help' for usage
        🔥 
```

/retest

**Comment by navidshaikh** [comment]
/retest


      result_collector.go:75: ERROR: Error: ReconcileIngressFailed: Ingress reconciliation failed

**Review by navidshaikh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
